### PR TITLE
feat(foundation): add file session recorder and checkpoint telemetry

### LINKED ISSUE #686: [Enhancement] Add FileSessionRecorder and checkpoint telemetry in StateGraph
## 🔍 Issue Description
Persistent debug sessions and checkpoint emissions are missing for the StateGraph path in `mofa-foundation`, which limits time-travel debugging and durable replay across process boundaries.

## 📌 Issue Type

- [ ] Bug
- [ ] Feature Request
- [x] Enhancement
- [ ] Documentation
- [ ] Refactor
- [ ] Other (please specify)

---

## 📝 Description
`SessionRecorder` has only in-memory support in foundation, and `GraphConfig.checkpoint_enabled/checkpoint_interval` were not producing telemetry events in `StateGraphImpl::invoke`.

- What is happening?
  - Sessions are not persisted to disk for StateGraph recorder usage.
  - Checkpoint configuration in runtime context has no emitted checkpoint events in StateGraph invoke flow.
- What should happen instead?
  - Add a file-backed recorder that persists session metadata/events.
  - Emit checkpoint telemetry events at configured step intervals when telemetry is attached.
- Why is this needed?
  - Durable session history and checkpoint telemetry are required for practical debugging, replay workflows, and future visual debugger UX.

---

## 🎯 Proposed Solution (Optional but Encouraged)
Implement in `crates/mofa-foundation`:

- Add `FileSessionRecorder` with JSONL format and configurable retention/size limits.
- Keep `SessionRecorder` trait compatibility and re-export recorder types in workflow module.
- Extend `CompiledGraphImpl` with optional `TelemetryEmitter` and emit checkpoint `DebugEvent::StateChange` events at `checkpoint_interval`.
- Add regression tests for recorder behavior and checkpoint interval emission.

Potential edge cases:
- Corrupt tail lines in JSONL files.
- Missing session files.
- Pruning policy should avoid active/running sessions.

---

## 📎 Additional Context
- Companion implementation branch: `feat/foundation-file-session-recorder-checkpoints`
- Preflight verification command:
  - `bash .codex/skills/mofa-pr-playbook/scripts/mofa_pr_preflight.sh --rust-targeted`

---

## 🙋 Claiming This Issue

To avoid duplicated work:

- [x] I'm willing to solve this issue by myself

### 🔔 Important
If you check the box above:
1. Comment below: **"I’d like to work on this."**
2. Wait for maintainer acknowledgment (if required).
3. Then start working.

I’d like to work on this.

### PR DESCRIPTION
**Pull request by Rahul-2k4**
## 📋 Summary
Add a file-backed session recorder for workflow debug telemetry and wire checkpoint event emission into `StateGraph` invocation when telemetry is attached.

## 🔗 Related Issues

Closes #686

Related to #334

---

## 🧠 Context
`SessionRecorder` persistence in foundation was limited to in-memory storage, and StateGraph checkpoint config fields were not producing telemetry emissions in the invoke loop. This change enables durable debug traces and explicit checkpoint signals without breaking existing APIs.

---

## 🛠️ Changes

- Added `FileSessionRecorder` and `FileSessionRecorderConfig` in `crates/mofa-foundation/src/workflow/session_recorder.rs` using JSONL session files.
- Added file-recorder tests covering round-trip, list/missing sessions, crash recovery, pruning, and concurrent session isolation.
- Extended `CompiledGraphImpl` with optional telemetry attachment via `with_telemetry(...)` and emitted checkpoint `DebugEvent::StateChange` events at configured intervals.
- Added `test_checkpoint_interval` to validate interval-based checkpoint emission counts.
- Re-exported `FileSessionRecorder` and `FileSessionRecorderConfig` from workflow module exports.

---

## 🧪 How you Tested

1. `bash .codex/skills/mofa-pr-playbook/scripts/mofa_pr_preflight.sh --rust-targeted`
2. Observed targeted crate detection for only:
   - `crates/mofa-foundation/src/workflow/mod.rs`
   - `crates/mofa-foundation/src/workflow/session_recorder.rs`
   - `crates/mofa-foundation/src/workflow/state_graph.rs`
3. Verified test results:
   - `cargo test -p mofa-foundation` => `240 passed; 0 failed`
   - doctests => `0 passed; 0 failed; 119 ignored`

---

## 📸 Screenshots / Logs (if applicable)
CLI/test logs are included in command output from preflight and local test runs.

---

## ⚠️ Breaking Changes

- [x] No breaking changes
- [ ] Breaking change (describe below)

If breaking:

---

### Code Quality
- [x] Code follows Rust idioms and project conventions
- [x] `cargo fmt` run
- [ ] `cargo clippy` passes without warnings

### Testing
- [x] Tests added/updated
- [x] `cargo test` passes locally without any error

### Documentation
- [x] Public APIs documented
- [ ] README / docs updated (if needed)

### PR Hygiene
- [x] PR is small and focused (one logical change)
- [x] Branch is up to date with `main`
- [x] No unrelated commits
- [x] Commit messages explain **why**, not only **what**

---

## 🚀 Deployment Notes (if applicable)
No migrations or env changes required.

---

## 🧩 Additional Notes for Reviewers
`cargo clippy -p mofa-foundation -- -D warnings` is currently blocked by a pre-existing unrelated lint in `crates/mofa-kernel/src/rag/types.rs` (`clippy::derivable_impls`) on the local checkout.

### CODE REVIEW

File: crates/mofa-foundation/src/workflow/state_graph.rs [hunk 1]
```diff
@@ -381,6 +391,28 @@ impl<S: GraphState + 'static> CompiledGraph<S> for CompiledGraphImpl<S> {
                 return Err(AgentError::Internal("Recursion limit reached".to_string()));
             }
             ctx.decrement_steps().await;
+            if let Some(emitter) = &self.telemetry
```
[the remark this answers is not in the document]
Comment (Rahul-2k4) [reply crates/mofa-foundation/src/workflow/state_graph.rs:547]: Fixed in 1e4a611: checkpoint emission now gates on `emitter.is_enabled()` in `CompiledGraphImpl::invoke`.

File: crates/mofa-foundation/src/workflow/session_recorder.rs [hunk 2]
```diff
@@ -49,6 +54,170 @@ impl Default for InMemorySessionRecorder {
...
+    fn session_path(&self, session_id: &str) -> PathBuf {
+        self.config
+            .session_dir
+            .join(format!("{}.jsonl", session_id))
+    }
+
+    fn session_tmp_path(&self, session_id: &str) -> PathBuf {
+        self.config
+            .session_dir
+            .join(format!("{}.jsonl.tmp", session_id))
```
[the remark this answers is not in the document]
Comment (Rahul-2k4) [reply crates/mofa-foundation/src/workflow/session_recorder.rs:142]: Fixed in 1e4a611: added strict `session_id` validation (only `[A-Za-z0-9_-]`) and return `InvalidInput` for unsafe IDs. Added test `test_file_recorder_rejects_invalid_session_id`.

File: crates/mofa-foundation/src/workflow/session_recorder.rs [hunk 3]
```diff
@@ -49,6 +54,170 @@ impl Default for InMemorySessionRecorder {
...
+            .drain(..)
+            .filter(|s| !active_ids.contains(&s.session_id) && s.status != "running")
+            .collect::<Vec<_>>();
+        completed.sort_by_key(|s| s.started_at);
+
+        let excess = self
+            .list_sessions()
+            .await?
+            .len()
+            .saturating_sub(self.config.max_sessions);
```
[the remark this answers is not in the document]
Comment (Rahul-2k4) [reply crates/mofa-foundation/src/workflow/session_recorder.rs]: Fixed in 1e4a611: `prune_sessions_if_needed` now computes `excess` from the first `sessions` snapshot (`total_sessions`) and no longer calls `list_sessions()` twice.

File: crates/mofa-foundation/src/workflow/session_recorder.rs [hunk 4]
```diff
@@ -105,6 +274,163 @@ impl SessionRecorder for InMemorySessionRecorder {
...
+        }
+
+        Self::append_json_line(
+            &path,
+            &SessionEventLine {
+                event: event.clone(),
+            },
+        )
+        .await?;
+
```
[the remark this answers is not in the document]
Comment (Rahul-2k4) [reply crates/mofa-foundation/src/workflow/session_recorder.rs:380]: Fixed in 1e4a611: added per-session async file locks (`file_locks` map + `Mutex`) and guard appends in `start_session`, `record_event`, and `end_session` to prevent interleaved JSONL writes.
